internal/reconciler/shoot/reconciler.go

```diff
@@ -23,10 +40,125 @@ type Reconciler struct {
-func (r *Reconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result, error) {
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	log := logf.FromContext(ctx)
 
-	log.Info("Shoot reconcile finished")
+	ctx, cancel := controllerutils.GetMainReconciliationContext(ctx, controllerutils.DefaultReconciliationTimeout)
+	defer cancel()
+
+	shoot := &gardencorev1beta1.Shoot{}
+	if err := r.Client.Get(ctx, req.NamespacedName, shoot); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(1).Info("Object is gone, stop reconciling")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("error retrieving shoot from store: %w", err)
+	}
+
+	if shoot.Annotations[v1beta1constants.AnnotationAuthenticationIssuer] != v1beta1constants.AnnotationAuthenticationIssuerManaged {
+		log.V(1).Info("Shoot does not have expected annotation or their value is not 'managed'", "annotation", v1beta1constants.AnnotationAuthenticationIssuer, "value", shoot.Annotations[v1beta1constants.AnnotationAuthenticationIssuer])
+		return reconcile.Result{}, nil
+	}
+
+	if shoot.Annotations[AnnotationTrustedShoot] != "true" && shoot.Annotations[AnnotationTrustedShoot] != "false" {
+		log.V(1).Info("Shoot does not have expected annotation or their value is not 'true'/'false'", "annotation", AnnotationTrustedShoot, "value", shoot.Annotations[AnnotationTrustedShoot])
+		return reconcile.Result{}, nil
+	}
+
+	if shoot.Annotations[AnnotationTrustedShoot] == "false" {
+		log.V(1).Info("Shoot has trusted annotation set to 'false', clean up OIDC resource")
+		return r.deleteOIDC(ctx, log, shoot)
+	}
+
+	if shoot.DeletionTimestamp != nil {
+		log.V(1).Info("Shoot is being deleted, clean up OIDC resource")
+		return r.deleteOIDC(ctx, log, shoot)
+	}
+
+	var issuerURL string
+	for _, adr := range shoot.Status.AdvertisedAddresses {
+		if adr.Name == v1beta1constants.AdvertisedAddressServiceAccountIssuer {
+			issuerURL = adr.URL
+			break
+		}
+	}
+
+	if issuerURL == "" {
+		log.Info("Shoot does not have service-account-issuer in its status.advertisedAddresses", "advertisedAddresses", shoot.Status.AdvertisedAddresses)
+		return ctrl.Result{RequeueAfter: r.ResyncPeriod}, nil
+	}
+
+	// TODO(theoddora): Add proper validation that a single issuer is not registered more than once
+	// This should check if another OIDC resource with the same issuerURL already exists
+
+	userNameClaim := "sub"
+	groupsClaim := "groups"
+	prefix := buildPrefix(shoot)
+	userNamePrefix := prefix
+	groupsPrefix := prefix
+
+	oidc := emptyOIDC(shoot)
+	if _, err := controllerutils.GetAndCreateOrMergePatch(ctx, r.Client, oidc, func() error {
+		oidc.Annotations = utils.MergeStringMaps(oidc.Annotations, map[string]string{})
+		oidc.Labels = utils.MergeStringMaps(oidc.Labels, map[string]string{
+			LabelKeyManagedBy: LabelValueGardenShootTrustConfigurator,
+		})
```

Comment:
I would vote to overwrite labels and annotations for now as I find it more practical and consistent. If there is a use case of allowing other labels and annotations on these resources we can think of applying a merge strategy.

```suggestion
		oidc.Annotations = nil
		oidc.Labels = map[string]string{
			LabelKeyManagedBy: LabelValueGardenShootTrustConfigurator,
		}
```